NeoML/src/Dnn/Dnn.cpp

```diff
@@ -497,6 +497,46 @@ void CDnn::ForceRebuild()
 	sourceLayers.SetSize( 0 );
 }
 
+CDnn* CDnn::CreateReferenceDnn(CRandom& random)
+{
+	NeoAssertMsg(referenceDnnRegistoror.referenceCounter != -1,
+		"resricted creating reference dnn if dnn is reference itself");
+	for(int i = 0; i < layers.Size(); ++i) {
+		auto* srcLayer = dynamic_cast<CSourceLayer*>(layers[i].Ptr());
+		if (srcLayer != nullptr) {
+			CheckArchitecture(srcLayer->GetBlob().Ptr() != nullptr, srcLayer->GetName(),
+				"initialize source blobs before creating reference dnns");
+		}
+	}
+
+	reshape();
+
+	CDnn* newDnn = new CDnn(random, mathEngine);
+	newDnn->DisableLearning();
+
+	for(int i = 0; i < layers.Size(); ++i) {
+		CPtr<CBaseLayer> copyLayer;
+		CMemoryFile file;
+		{
+			CArchive archive(&file, CArchive::SD_Storing);
```

Comment:
How will it be faster?
1
```
CMemoryFile file;
for(int i = 0; i < layers.Size(); ++i) {
	{
		CArchive archive(&file, CArchive::store);
		CPtr<CBaseLayer> originalPtr(layers[i]);
		SerializeLayer(archive, layers[i]->MathEngine(), originalPtr);
	}
	file.SeekToBegin();
	CPtr<CBaseLayer> copyLayer;
	{
		CArchive archive(&file, CArchive::load);
		SerializeLayer(archive, layers[i]->MathEngine(), copyLayer);
		layers[i]->transferParamsBlob(*copyLayer);
	}
	newDnn->AddLayer(*copyLayer);
}
```

or

2
```
CMemoryFile file;
{
	CArchive archive(&file, CArchive::store);
	for(int i = 0; i < layers.Size(); ++i) {
		SerializeLayer(archive, layers[i]->MathEngine(), layers[i]);
	}
}
file.SeekToBegin();
{
	CArchive archive(&file, CArchive::load);
	CPtr<CBaseLayer> copyLayer;
	for(int i = 0; i < layers.Size(); ++i) {
		SerializeLayer(archive, layers[i]->MathEngine(), copyLayer);
		layers[i]->transferParamsBlob(*copyLayer);
		newDnn->AddLayer(*copyLayer);
	}
}
```

Could you please check?

Reply:
First one is faster, but "CMemoryFile file" need to be in cycle (we will read first layer data every time in cycle otherwise)

Reply:
ok, will use first